### DISCUSSION TITLE
Any way to elevate permissions of postgres user?

### DISCUSSION THREAD

**Discussion by leesus** [supabase/supabase#1535] [Questions] [answered]
Is there a way to grant `superuser` role to `postgres` user? I'm using prisma to connect to the database, but while some of my migrations work locally against the docker image, they fail against supabase-hosted databases due to permissions errors.

Is there anyway to do this?

**Comment by leesus** [comment, marked as answer]
I just ran a script in the UI. I should probably have tried it out before asking the question!

**Reply by inian** [reply]
Hi @leesus, may I ask what your use case is that requires postgres to be a superuser? We are going to be restricting it from being a superuser (even from the dashboard).

**Reply by jasonho-lynx** [reply]
@inian can I confirm if this restriction has already happened and will be the case going forward? I noticed that in a more recent db instance, postgres is no longer a superuser, unlike that in an older instance

**Reply by inian** [reply]
Hey @jasonho-lynx, yup for new projects postgres is no longer a superuser.

**Reply by quicksnap** [reply]
@inian Is Supabase going to be disallowing promoting to superuser? I enjoy the freedom.. :)

**Reply by inian** [reply]
Haha @quicksnap :) The problem is that there are some roles (anon, authenticated, etc) and tables (in storage and auth schema) we expect to be present for Supabase to function properly. Changing these can break upgrades and existing functionality. Plus there are powerful postgres commands available to the superuser which can be misused (like installing crypto miners). This is the reason why we had to demote postgres. 

However if there is some specific usecase that requires postgres to be a superuser, let me know. We definitely do not want to restrict what you can do with the underlying database if possible.

**Reply by cdedreuille** [reply]
I just wanted to add to the discussion here. In my case, I just wanted to delete an auth user from the dashboard but I couldn't because I was not a superuser. I temporarily put Postgres to superuser and it worked. Not sure what's the best way to handle it though.

**Reply by inian** [reply]
Hey @cdedreuille, deleting auth users from the dashboard (/https://app.supabase.io/project/project-ref/auth/users) should still work. If this isn't working, let me know!

**Reply by cdedreuille** [reply]
@inian I changed back postgres to non super user and trying to delete a user from the dashboard I get this message:

<img width="382" alt="Screenshot 2021-05-31 at 14 46 18" src="https://user-images.githubusercontent.com/1540635/120195378-0181f180-c21f-11eb-9a62-2245ededb419.png">

**Reply by inian** [reply]
Hmm am not able to replicate this on a new project Charles (where postgres shouldn't be a superuser). Any chance you altered permissions that is specific to your project? If not, open an issue and lets discuss there!

**Reply by d-mok** [reply]
I have a situation where promoting `postgres` to `superuser` seems to be required.
I use a 3rd party service to backup the supabase database. I give it the `postgres` credential for that. When `postgres` is `superuser`, the backup works. But when `postgres` is not `superuser`, the backup encounter errors like:
```
pg_dump: error: query failed: ERROR:  permission denied for sequence refresh_tokens_id_seq
pg_dump: error: query was: SELECT last_value, is_called FROM auth.refresh_tokens_id_seq
```
It seems that `postgres` doesn't have enough permission for a backup job if it's not `superuser`. However, I know that a backup job should definitely NOT require a `superuser` role to work properly. So, is there any other way to work around this? What permission should I (or can I) give to `postgres` for my backup job to work?

**Reply by jasonho-lynx** [reply]
@d-mok a backup job indeed does not require superuser access, but errors will occur if person executing the backup (`postgres` in this case) does not have access to some of the database objects (`auth.refresh_tokens_id_seq` in this case). 

Granting superuser to `postgres` is one good ol' blast-it-all solution since it grants access to everything, but it's probably more than what you need. A better idea is to specifically grant access to those objects to the user executing the backup. A sample solution for your example can be found here: https://stackoverflow.com/questions/9325017/error-permission-denied-for-sequence-cities-id-seq-using-postgres, where access to the sequence is manually granted. The same will have to be done for any other objects (tables, functions, etc.) that cause errors while being backed up.
